Add remote MCP server docs (#176)
superfly/sprites-docs: src/content/docs/integrations/remote-mcp.mdx

---
title: Remote MCP Server
description: Connect MCP clients to Sprites with the hosted remote MCP server
---

import { Callout } from '@/components/react';

> Sprites provides a hosted remote MCP server. Once connected, an MCP client can create and manage Sprites, run commands inside them, inspect checkpoints and services, and adjust network policy through standard MCP tools.

Reviewer: ```suggestion
Sprites runs a hosted MCP server. Connect an MCP client to it and the client can control Sprites directly. It can create and manage them, run commands, inspect checkpoints, and change network policy through standard MCP tools.
```

> Use it when you want an MCP client to work in isolated, persistent Linux environments without copying API tokens into prompts or running a local MCP bridge.

Reviewer: ```suggestion
Why use it? Your MCP client gets persistent, isolated Linux environments without you pasting API tokens into a prompt or running a local MCP bridge.
```

## MCP server URL

Use this URL when adding the connector:

```text
https://sprites.dev/mcp
```

The connector uses OAuth. During setup, you sign in with your Fly.io account, choose the organization the client should use, and approve the level of Sprites access for that connector.

## Add the connector

1. Open your MCP client's connector settings.

> 2. Add a custom or remote MCP connector.

Reviewer: ```suggestion
2. Add a remote (or custom) MCP connector. Different clients label this differently. Look for "Remote MCP server", "Custom connector", or similar.
```

3. Enter `https://sprites.dev/mcp`.

> 4. Follow the browser authentication flow.

Reviewer: ```suggestion
4. Follow the browser authentication flow for Fly.io
```

> 4. Follow the browser authentication flow for Fly.io

Reviewer: ```suggestion
4. Follow the browser authentication flow for Fly.io.
```

5. Choose the Sprites organization the connector should use.
6. Review the token restrictions on the consent screen and approve the connection.

<Callout type="info">
If your client asks for a connector name, use `Sprites`.
</Callout>

## Permissions

> By default, the OAuth consent flow creates a restricted Sprites token for the MCP connector. The default restriction allows the connector to create Sprites with the `mcp-` name prefix and limits the number of Sprites it can create.

Reviewer: ```suggestion
The default token is restricted. The connector can only create Sprites whose names start with mcp-, and it can only create a limited number of them.
```

> The default token is restricted. The connector can only create Sprites whose names start with mcp-, and it can only create a limited number of them.

Reviewer: ```suggestion
The default token is restricted. The connector can only create Sprites whose names start with `mcp-`, and it can only create a limited number of them.
```

> During consent, you can adjust the name prefix and maximum Sprite count. You can also choose full token access if you want the connector to operate on the organization without those MCP-specific restrictions.

Reviewer: ```suggestion
You can change both on the consent screen: pick a different name prefix, raise or lower the cap, or switch to full access if you want the connector to act on the whole organization without those guardrails.
```

<Callout type="warning">
Full access lets the connector manage any Sprite the selected organization token can access. Use restricted mode unless your workflow needs broader control.
</Callout>

## Available tools

> MCP clients see Sprites as tools. The hosted server exposes organization-level tools and Sprite-level tools.

Reviewer: ```suggestion
MCP clients see Sprites as tools. There are two layers: organization-level tools that act across your Sprites, and Sprite-level tools that act inside a specific one.
```

| Tool | Access | What it does |
|------|--------|--------------|
| `list_sprites` | Read-only | Lists Sprites visible to the authenticated organization and token policy. |
| `create_sprite` | Destructive | Creates a Sprite in the selected organization. |
| `destroy_sprite` | Destructive | Permanently destroys a Sprite. |

### Sprite tools

The Sprite-level tools are generated from the public Sprite environment API. The exact list may change as new Sprite environment versions ship, but commonly includes:

| Category | Examples |
|----------|----------|
| Exec | Run a command, list exec sessions, kill a session |
| Checkpoints | Create, list, inspect, and restore checkpoints |
| Network policy | Read or update outbound network rules |
| Services | List, create, start, stop, and inspect background services |
| Service logs | Read recent service logs |

Tools are annotated as either read-only or destructive. Your MCP client may ask for additional confirmation before using destructive tools, depending on its settings.

## Safety model

Sprites are isolated Linux environments. That isolation is useful for agent work, but MCP tools can still perform real actions in your organization.

> - `destroy_sprite` permanently deletes a Sprite and its data.

Reviewer: ```suggestion
- `destroy_sprite` permanently deletes a Sprite and its data. There is no undo.
```

- Command execution can modify files, install packages, start services, or use network access allowed by the Sprite policy.

> - Checkpoint restore changes a Sprite back to a previous filesystem state.

Reviewer: ```suggestion
- Checkpoint restore rewinds a Sprite to a previous filesystem state. Any work done since that checkpoint is gone.
```

- Network policy updates can allow or block outbound access from a Sprite.

For safer workflows, use a restricted OAuth token, pick a clear name prefix such as `mcp-`, and create a fresh Sprite for experiments.

## Example requests

```text
Create a Sprite named mcp-demo, run python --version, and tell me what runtime is installed.
```

```text
Create a checkpoint on mcp-demo before making changes. Then install ripgrep and verify rg --version.
```

```text
List my Sprites with the mcp- prefix and summarize which ones are running.
```

## Troubleshooting

### The client cannot connect

Check that the connector URL is exactly:

```text
https://sprites.dev/mcp
```

The MCP endpoint is not the REST API host. Do not use `https://api.sprites.dev`.

### Authentication loops or fails

> Sign out of Fly.io in your browser, sign in again, and retry the connector flow. If you belong to multiple organizations, make sure you select the organization that should own Sprites created through MCP.

Reviewer: ```suggestion
Sign out of Fly.io in your browser, sign in again, and retry the connector flow. If you belong to multiple organizations, make sure you select the one that should own Sprites created through MCP.
```
